feat(route)path regex matches support req_uri with args, disable route cache

mnteam opened the pull request:
### Summary
add path regex matches support request uri with args, disable route cache for head mapping with uri issue
refer the feature link: https://github.com/Kong/kong/issues/5301

### Full changelog

/kong/router.lua

### Issues resolved
1. Add path regex matches support request uri with args
2. disable cache for route matches to avoid the header route issue.

All above changes tested(including stress testing) and the result is OK.

For point 2, I need to explain as below: a simple case, Please tell me if I get wrong, thank you very much.

there are two routing with different service:
route a: path: /api/xxx/one, with head matches: x-token=foo
route b: path: /api/xxx/one

while any request match a, then the rest of request take the cache logic with below code:

``` lua
-- cache lookup (except for headers-matched Routes)
    local cache_key = req_method .. "|" .. req_uri .. "|" .. req_host ..
                      "|" .. ctx.src_ip .. "|" .. ctx.src_port ..
                      "|" .. ctx.dst_ip .. "|" .. ctx.dst_port ..
                      "|" .. ctx.sni
    do
      local match_t = cache:get(cache_key)
      if match_t then
        return match_t
      end
    end
```

NOTE: the general way, last five args in cache key is empty. which means the cache key is decided by req_method, req_uri and req_host.

which means the requests which should be matched with route b, but by this cache logic, it would match the cache a(no change to check the head parts).

Another point, involve the args with request uri support, we should not support cache, otherwise, the cache would have a low hit rate.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=5347) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=5347) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=5347) it.</sub>

dndx commented:
Hello @mnteam,

Thanks for your PR! I have some thoughts on this functionality after reading it.

1. I am worried doing regex matching on the full path (with arguments) is a dangerous approach. Consider the following scenario:
   * `/foo?bar=1&baz=2`
   * `/foo?baz=2&bar=1`

    Semantically those two paths are identical, but it can not be easily represented using regex. This seems to 
    create more confusion for the user than it's worth.

    What I believe may be a better approach is to follow what the header based matching PR 
    (https://github.com/Kong/kong/pull/4758) did and make sure we are matching the query arguments by 
    *semantic* instead of *string*.

2. Not sure if I got you correctly but right now if header based matcher is evaluated, the match result is never cached. Please see: https://github.com/Kong/kong/blob/ff8b4cf/kong/router.lua#L1647-L1649

    To make sure of that I created the scenario you described locally and wasn't able to reproduce any cache 
    poisoning issues. Would you mind share your reproducible configuration with me?

Thanks

mnteam commented:
Hi,

Thanks for your response, 

For 1, actually we use only one parameter pair for routing, in your case, that is actually the issue, I added it because it is supported by nginx, and we replaced from nginx to kong, to avoid more changes by biz side, maybe this feature is not required for new
 structure. I agree with you, if there are multiple arguments to matches, it is better to use header value pair.

For 2, I checked the latest code, it has been fixed, refer the key point of codes: router.lua -> line[1369-1385, 1387]. So you can ignore this issue right now.
(In our scenario, we have to disable all cache, because the req_uri with args support would be out of control with cache).

Anyway thanks very much for supporting.

Best Regards,
Peyton

> 在 2020年1月14日，上午8:50，Datong Sun <notifications@github.com> 写道：
> 
> Hello @mnteam <https://github.com/mnteam>,
> 
> Thanks for your PR! I have some thoughts on this functionality after reading it.
> 
> I am worried doing regex matching on the full path (with arguments) is a dangerous approach. Consider the following scenario:
> 
> /foo?bar=1&baz=2
> /foo?baz=2&bar=1
> Semantically those two paths are identical, but it can not be easily represented using regex. This seems to
> create more confusion for the user than it's worth.
> 
> What I believe may be a better approach is to follow what the header based matching PR
> (#4758 <https://github.com/Kong/kong/pull/4758>) did and make sure we are matching the query arguments by
> semantic instead of string.
> 
> Not sure if I got you correctly but right now if header based matcher is evaluated, the match result is never cached. Please see: https://github.com/Kong/kong/blob/ff8b4cf/kong/router.lua#L1647-L1649 <https://github.com/Kong/kong/blob/ff8b4cf/kong/router.lua#L1647-L1649>
> To make sure of that I created the scenario you described locally and wasn't able to reproduce any cache
> poisoning issues. Would you mind share your reproducible configuration with me?
> 
> Thanks
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Kong/kong/pull/5347?email_source=notifications&email_token=ABCYB2SLFPBHBJHBTL2GRUTQ5UD4FA5CNFSM4J4ETFJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI23M2A#issuecomment-573945448>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABCYB2R6JLBSQHBI5XC4OSDQ5UD4FANCNFSM4J4ETFJQ>.
>

dndx commented:
Hi @mnteam,

Yes, for an PR to be considered into the official Kong repository it has to solve some common use cases among Kong users. If you can rewrite this PR so that it follows a similar approach as the route by header PR I think it would meet that requirement.

Also, some quick test cases would be nice. Again, #4758 is a good reference. Let me know if you need any help along the way!

Thanks

hishamhm commented:
Closing this one as we agree it's not the way to go, if alternative approaches are considered, let's do them in a new PR!
